### PR TITLE
Chem dispenser supports negative macro values according to macroresolution

### PR DESCRIPTION
**Pull request by yorii**
[Guidelines]: # (Be sure that your PR follows our guidelines, such as modularization and comment standards. You can read more about the subject here: https://github.com/HippieStation/HippieStation/blob/master/hippiestation/README.md )

## Changelog
:cl: yorii
tweak: Chem dispenser now supports negative values in macros.
/:cl:

## About The Pull Request 
Adds a quick and easy QoL improvement/buff for chemists who like to write more advanced macros.

## Why It's Good For The Game
It saves you a few clicks and from future carpal tunnel syndrome.

### CODE REVIEW

**Comment by ghost** [comment]
+1 but i'd personally prefer it be all tiers, but with precision based on tier

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 1]
```diff
@@ -233,10 +233,16 @@
...
 					say("Not enough energy to complete operation!")
 					return
-				R.add_reagent(reagent, actual)
+				if(actual > 0)
+					R.add_reagent(r_id, actual)
+				if(actual < 0)
+					if(macroresolution == 1)
+						R.remove_reagent(r_id, abs(actual))
+					if(macroresolution > 1)
+						say("Femtomanipulator required to remove chemicals from mixture!"
```
Comment (ghost) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: you forgot a `)`

**Comment by yorii** [comment]
This commit was created on GitHub.com and signed with a verified signature using GitHub’s key.

You can sue me if this code doesn't work.

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm (hunk 1 above)
Comment (yorii) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: FUGGGGGGGGGGg..... I am commit sudoku now

**Comment by yorii** [comment]
> +1 but i'd personally prefer it be all tiers, but with precision based on tier

Oh... well in that case...

**Comment by yorii** [comment]
it was even less code changes then :3

**Review by ghost — CHANGES_REQUESTED** [review]
YOU DIDN'T EVEN DO THIS RIGHT -- you edited the `dispense` topic, not `dispense_recipe`

LUL

**Comment by yorii** [comment]
it's the same isn't it? the recipe calls on this function to actually do the dispensing?

**Comment by yorii** [comment]
oh no shit you're right you have it split up for two separate add_reagent functions...

**Comment by yorii** [comment]
You forgot to adjust the powerusage for negative values, unless we want removing reagents to recharge the powercell? (which I suspect will allow for some infinite recharge shenanigan macros)

EDIT: nvm that's not even done there at all, but it still needs to check for the abs(actual) value otherwise it won't want to accept negative actual at all.

**Review by yorii — COMMENTED** [review]
is all good now

**Review by ghost — DISMISSED** [review]

**Comment by yorii** [comment]
Hi sorry, these changes aren't enough it seems, only tested them on tg until now and you seem to have a slightly different chem_dispenser.dm, don't merge for now :)

I think it's this part:
`max(chemicals_to_dispense[key], res)`

**Comment by yorii** [comment]
Changed a max to a round and another missing abs(value) for the recipe validation check, but now it's actually even tested 🥇

**Review by ghost — APPROVED** [review]

**Comment by cris5311** [comment]
+1


Yeah thats good shit, now we only need super macros which can call on saved macros

**Comment by Manly-Dwarf** [comment]
chemistry powergame never ends
time to make entire recipes in one click

**Comment by DONKARMA** [comment]
why

**Comment by Manly-Dwarf** [comment]
why

**Pull request merged**